### PR TITLE
[REVIEW] Implement DataFrame.__setattr__ to enable setting items

### PR DESCRIPTION
**Pull request by cwharris**
See #1963

There seemed to be a couple of options to implement `__setattr__` without interfering with instance-level attributes. The first was to create a list which declared all attributes, treating them as attributes, whereas all other keys/names would be forwarded to `__setitem__`. The second approach was to assume all attributes would be established prior to leaving `__init__`. Past that point, any key which did not correspond to an existing attribute will be instead treated as a column.

This PR represents the latter approach, which I chose because no properties were being set after `__init__`, and the declarative style does not require a list of attributes, which seems redundant. I can see that if we wanted to add "future" reserved words, a list would make sense.

*TLDR;*
Ability to set columns via attribute/property syntax, given that attribute has not already been established. Assumes all attributes will be assigned prior to leaving `__init__`. Assigning _new_ attributes prior to leaving `__init__` is not possible with this approach.

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -164,6 +164,8 @@ def __init__(self, data=None, index=None, columns=None):
 
         self._add_empty_columns(columns, index)
 
+        self._enable_setattr_setitem = True
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py]: Is there ever a situation we want this set to False?

File: python/cudf/cudf/tests/test_dataframe.py [hunk 2]
```diff
@@ -226,14 +226,16 @@ def test_dataframe_column_add_drop():
     df = DataFrame()
     data = np.asarray(range(10))
     df["a"] = data
-    df["b"] = data
+    df.b = data
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py]: This shouldn't work. This should throw a warning similar to:
```
/home/nfs/kkraus/miniconda3/envs/cudf_dev/bin/ipython:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
```

You should only be able to change data via a `__setattr__` call.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/core/dataframe.py]: In the initializer prior to adding, for example, private attributes. Otherwise you couldn’t set them.

Comment (cwharris) [reply python/cudf/cudf/core/dataframe.py]: The alternative is “registering” all attributes explicitly, and checking that list inside the dunder.

**Comment by thomcom** [comment]
Cool and parsimonious solution to a tricky problem! I looked at this for a day and did not come up with this solution, nice.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (thomcom) [reply python/cudf/cudf/core/dataframe.py]: This is a runtime flag - basically it says that `__init__` has completed, and future `__setattr__` attempts will have to pass through the new logic.

Should it be named `_init_complete` or similar to reflect that?

Comment (thomcom) [reply python/cudf/cudf/core/dataframe.py]: I'm fairly opposed to the whitelist solution, so I like this.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 2 above)
Comment (thomcom) [reply python/cudf/cudf/tests/test_dataframe.py]: I think you need substantially more tests. This logic is -very-high-risk- in that it affects every __attr__ access to a dataframe. I'd like to see tests that:

1. attempt to do what Keith has mentioned here and throw an error ala Pandas
2. valid re-assignments
3. re-re-assignments
4. attempting to `__setattr__` on methods throwing an error
5. attempting to `__setattr__` a valid attribute with invalid data
6. `__setattr__` at the same time as `__getattr__`, can you make them collide?

There are more corner cases here that aren't immediately hopping to mind.

Comment (thomcom) [reply python/cudf/cudf/tests/test_dataframe.py]: Another important test:
7. `__setattr__` on a copy of an object, validating that the `__init__` control works properly.
8. There are ways to copy objects without calling `__init__` on them, we need to discover that case and handle it properly.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/core/dataframe.py]: I was thinking about how we could create new attributes after-the-fact even when the names correspond to an existing column, and thought that the context management protocol would work well. Something like...

```
df["a"] = [0, 1, 2]
with df.attr_lock:
  df.a = "attribute"

print(df.a) # prints "attribute"
```
where `attr_lock` is a property which returns an object supporting the context management protocol. Under the hood it was just disable and re-enable the `_enable_setattr_setitem` flag.

Just a thought. Not looking to add something like that right now.

File: python/cudf/cudf/core/dataframe.py [hunk 3]
```diff
@@ -269,6 +271,21 @@ def __dir__(self):
...
+        try:
+            object.__getattribute__(self, key)
+            return object.__setattr__(self, key, col)
+        except AttributeError:
+            pass
+
+        try:
+            if self._enable_setattr_setitem:
+                self.__setitem__(key, col)
+        except AttributeError:
```
Comment (cwharris) [inline python/cudf/cudf/core/dataframe.py]: I had a warning here. I'm not sure what happened to it. I'll get it back in there.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 2 above)
Comment (cwharris) [reply python/cudf/cudf/tests/test_dataframe.py]: Agreed that we definitely need more substantial tests. I'm curious about the object copy scenario. It seems unlikely that a dataframe would be copied in such a manner, or that someone would want to add attributes after doing so. It's a duck-typed language, so lots of weird things are possible. Is there precedence for supporting this scenario?

Comment (cwharris) [reply python/cudf/cudf/tests/test_dataframe.py]: Pandas actually issues a warning. I had one going last night, but apparently I deleted it. (I shouldn't have stayed up so late). I'll get the error back in there.

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/core/dataframe.py]: I thought about naming the flag `_init_complete`, but decided against it because it was less specific, and assumed the flag would never be unset.

File: python/cudf/cudf/core/dataframe.py [hunk 4]
```diff
@@ -163,6 +163,7 @@ def __init__(self, data=None, index=None, columns=None):
                 self.add_column(col_name, series, forceindex=index is not None)
 
         self._add_empty_columns(columns, index)
+        self._allow_setattr_to_setitem = True
```
Comment (thomcom) [inline python/cudf/cudf/core/dataframe.py]:
```suggestion
        # this flag marks the completion of `__init__`, which changes the behavior
        # of `__setattr__` to enable the Pandas-like attr functionality `df.x = col`.
        self._allow_setattr_to_setitem = True
```

Comment (cwharris) [reply python/cudf/cudf/core/dataframe.py]: IMO `# allows Pandas-like __setattr__ functionality: 'df.x = column', etc.` is more appropriate, since it's possible to use this flag after `__init__` to patch in new attributes without worrying about overlapping column names.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 2 above)
Comment (cwharris) [reply python/cudf/cudf/tests/test_dataframe.py]: @thomcom I've addressed some of your concerns, but I don't see why others need tests (and therefore don't know how to test).

I've added assignment + reassignment loop test, and warning on `__setattr__` for new attributes.

I'm confused how `__setattr__` and `__getattr__` could ever collide, given the GIL, and that this is pure python.
I'm also confused why attempting to __setattr__ on a method would raise that method's error, given that python doesn't invoke it, and our dunder override doesn't either.
Also, since we're forwarding to `__setitem__`, I'm not sure it it make sense to re-test things such as setting a valid attribute with invalid data? (assuming by valid attribute you mean column, otherwise anything is valid)

**Comment by harrism** [comment]
Gave this PR a more descriptive title.

**Comment by cwharris** [comment]
Giving @thomcom a chance to re-review before merging.

**Comment by thomcom** [comment]
Thanks for letting me take a look! I'll post approval pretty soon...
